### ISSUE TITLE
Discuss possible code checks to be used in meshpy

### ISSUE THREAD

**Issue by isteinbrecher** [imcs-compsim/meshpy#108]
This issue shall serve as a spot to discuss possible code checks and pre-commit hooks to be used within MeshPy. Some possible choices can be seen here https://github.com/davidrudlstorfer/pyskel/blob/main/.pre-commit-config.yaml (thanks to @davidrudlstorfer).

Possbile tools:
- [x] `pylint` currently we have `8.93/10` for `meshpy/` and `8.55/10` for `tests/`
- [x] Ruff
- [x] ~Code formatter for markdown~ could not be found
- [x] Code formatter for c++
- [x] ~Code formatter for Cython. Does such a tool exist? Do we need this as the Cython code likely will not change on a regular basis.~ Deemed not worth the effort
- [x] Typo check (see #163)

Other code format to enforce:
- [x] Require docstrings for functions
- [ ] Permit relative imports depends on #159

**Comment by davidrudlstorfer** [comment]
To note it for later discussion; we could also utilize ruff (https://github.com/astral-sh/ruff) which replaces flake8, isort and black in one tool. This tool is currently spreading fast

**Comment by isteinbrecher** [comment]
I think with #153 we got a very long way towards completing this issue. Are there still some things missing? Two I can think of are the required docstrings and prohibiting relative imports (I added them to the description).

**Comment by davidrudlstorfer** [comment]
#153 includes a first suite of pre-commit hooks and also utilizes it within the code check pipeline

Further points to complete code checks are:

- Enforce absolute imports, i.e., meshpy.rotation.blablub rather than ..rotation.blablub
- Enforce the docstrings, currently we have 96% coverage => add remaining docstrings and set 100% coverage
- Enforce the docstring style, i.e. args to be consistent across the code base
- Type hinting - I think we briefly talked about this and you had some points against that?
- Enforce that each source code file contains the license header

**Comment by davidrudlstorfer** [comment]
- Run xml formatting and syntax check on vtu/vtk files (per https://github.com/imcs-compsim/meshpy/pull/153#issuecomment-2545871930)

**Comment by davidrudlstorfer** [comment]
Regarding the xml formatting and syntax check on vtu/vtk files:

Yesterday I wasted a lot of time to get this running and I was not able to get it to work at all. I was not able to add differing file endings to the xml hooks.

We probably need our own special hook for that. Maybe we can just take the contents of the vtu/vtk files and hand it over to the xml hooks.

**Comment by isteinbrecher** [comment]
That was just a suggestion, we don't have to do that if it is complicated. The relevant files are almost never created by hand so I am fine with just sticking with the default format there (now that I think of it, I think in 4C we also don't format those files - at least I never had any issues checking in the automatically created ones).

**Comment by isteinbrecher** [comment]
@davidrudlstorfer is it possible to automatically format standard comments (after a `#`) similar to docstrings?

**Comment by davidrudlstorfer** [comment]
I recently looked a bit into auto formatting comments but was not able to find any specific pre-commit hooks which could do that.

The only related thing I found was https://github.com/astral-sh/ruff/issues/7414

So once ruff supports it we can simply turn it on in out pre-commit hook

In my opinion we can soon close this issue as all possible code checks are in place after #167 is merged. I would propose to add the TODO for the relative imports in the issue #169

**Comment by isteinbrecher** [comment]
I agree, let's close this issue as the "initial" pre-commit wave is completed and open individual issues for new hooks in the future. I will move the point regarding the relative imports to #159
